src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+        .func_wrap("env", "font_glyph_to_string", font_glyph_to_string)
+        .ok()?
+        .func_wrap("env", "font_get_glyph_h_advance", font_get_glyph_h_advance)
+        .ok()?
+        .func_wrap("env", "font_get_glyph_v_advance", font_get_glyph_v_advance)
+        .ok()?
+        .func_wrap("env", "font_copy_glyph_outline", font_copy_glyph_outline)
+        .ok()?
+        .func_wrap("env", "face_copy_table", face_copy_table)
+        .ok()?
+        .func_wrap("env", "buffer_copy_contents", buffer_copy_contents)
+        .ok()?
+        .func_wrap("env", "buffer_set_contents", buffer_set_contents)
+        .ok()?
+        .func_wrap("env", "debugprint", debugprint)
+        .ok()?
+        .func_wrap("env", "shape_with", shape_with)
+        .ok()?;
+
+    // Here we are (supposedly) done creating functions.
+    // draft section
+
+    // The WASM code inside a font is expected to export a function called shape which takes five int32 arguments
+    // and returns an int32 status value. (Zero for failure, any other value for success.) Three of the five
+    // arguments are tokens which can be passed to the API functions exported to your WASM code by the host
+    // shaping engine:
+    //
+    // A shape plan token, which can largely be ignored.
+    // A font token.
+    // A buffer token.
+    // A feature array.
+    // The number of features.
+
+    let instance = linker.instantiate(&mut store, &module).ok()?;
+    let shape = instance
+        .get_typed_func::<(u32, u32, u32, u32, u32), i32>(&mut store, "shape")
+        .ok()?;
+
+    if let Ok(0) | Err(_) = shape.call(&mut store, (0, 0, 0, 0, 0)) {
+        return None;
+    };
+
+    let ret = store.into_data().buffer;
+
+    Some(GlyphBuffer(ret))
+}
+
+// ===========
+// functions imported  into WASM
+// ===========
+
+// fn face_get_upem(face: u32) -> u32;
+// Returns the units-per-em of the font face.
+fn face_get_upem(caller: Caller<'_, ShapingData>, _face: u32) -> u32 {
+    caller.data().font.units_per_em as u32
+}
+
+// fn font_get_face(font: u32) -> u32;
+// Creates a new face token from the given font token.
+fn font_get_face(caller: Caller<'_, ShapingData>, _font: u32) -> u32 {
+    // From HarfBuzz docs:
+    // (In the following functions, a font is a specific instantiation of a face at a
+    // particular scale factor and variation position.)
+    //
+    // I am unsure how to represent that in rustybuzz.
+    // er .. do what here?
+    0
+}
+
+// fn font_get_glyph(font: u32, unicode: u32, uvs: u32) -> u32;
+// Returns the nominal glyph ID for the given codepoint, using the cmap table of the font to map Unicode codepoint (and variation selector) to glyph ID.
+fn font_get_glyph(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    codepoint: u32,
+    uvs: u32,
+) -> u32 {
+    let Some(codepoint) = char::from_u32(codepoint) else {
+        return 0;
+    };
+
+    if uvs == 0 {
+        caller.data().font.glyph_index(codepoint).unwrap().0 as u32
```

Comment:
should we also return 0 here instead of unwrapping?

Reply:
Yes that was me debugging the function. Already fixed locally.